File: apps/android/app/src/main/kotlin/dev/typvia/mobile/UiPreferences.kt
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

package dev.typvia.mobile

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import dev.typvia.mobile.ui.UiPreference
import dev.typvia.mobile.ui.UiPreferenceFile
import java.io.File

/**
 * Where this device keeps the reader's two choices about how the app looks.
 *
 * Two strings, read before the window opens and written the moment they
 * change. Nothing about snippets is kept here and nothing here is ever synced:
 * these are facts about one screen, not about a library.
 *
 * The store is the platform's own rather than a database or a file of our own —
 * the values are two words, they must be readable synchronously on the way to
 * the first frame, and the framework already keeps that kind of thing.
 */
class UiPreferences internal constructor(
    private val store: SharedPreferences?,
    /**
     * The package's data directory, where the surfaces that run in their own
     * process look. Absent only where the platform would not name it, and
     * then this app is simply the only one that knows the reader's choice.
     */
    private val dataDir: File?,
) {
    private var languageState by mutableStateOf(
        UiPreference.Language.of(read(Keys.LANGUAGE)),
    )
    private var appearanceState by mutableStateOf(
        UiPreference.Appearance.of(read(Keys.APPEARANCE)),
    )

    var language: UiPreference.Language
        get() = languageState
        set(value) {
            languageState = value
            write(Keys.LANGUAGE, value.stored)
            publish()
        }

    var appearance: UiPreference.Appearance
        get() = appearanceState
        set(value) {
            appearanceState = value
            write(Keys.APPEARANCE, value.stored)
            publish()
        }

    /**
     * Puts the choice where the other processes look.
     *
     * The keyboard and the share sheet are not in this composition and cannot
     * be told; they read a file. Written on every change rather than on a
     * timer, because the reader expects the keyboard they pull up next to be
     * the one they just chose.
     */
    private fun publish() {
        dataDir?.let {
            UiPreferenceFile.write(
                it,
                UiPreferenceFile.Chosen(languageState, appearanceState),
            )
        }
    }

    private fun read(key: String): String? = runCatching { store?.getString(key, null) }.getOrNull()

    private fun write(key: String, value: String) {
        // A preference that could not be stored still applies to this run: the
        // reader chose it, and refusing to render their choice because it could
        // not be remembered would be losing twice instead of once.
        runCatching { store?.edit()?.putString(key, value)?.apply() }
    }

    /**
     * The keys are the ones the desktop and the other phone use, so the three
     * ends read as one product to whoever goes looking.
     */
    /**
     * Writes the file only where it does not already say this.
     *
     * A first run after an update has a stored preference and no file yet, and
     * a reader who never opens this app again would otherwise never get their
     * choice across to the keyboard. Reading two lines on the way to the first
     * frame is cheaper than the store this class already opened there.
     */
    private fun publishIfStale() {
        val directory = dataDir ?: return
        val wanted = UiPreferenceFile.Chosen(languageState, appearanceState)
        if (UiPreferenceFile.read(directory) != wanted) UiPreferenceFile.write(directory, wanted)
    }

    internal object Keys {
        const val LANGUAGE = "tv.ui.locale"
        const val APPEARANCE = "tv.ui.theme"
    }

    companion object {
        private const val FILE = "tv.ui"

        /**
         * Opens the store, or carries on without one.
         *
         * A device whose preference file cannot be opened is a device that
         * follows the system and forgets what it was told — which is the state
         * every device starts in anyway. It is not a reason to fail to start.
         */
        fun open(context: Context): UiPreferences = UiPreferences(
            runCatching { context.getSharedPreferences(FILE, Context.MODE_PRIVATE) }.getOrNull(),
            runCatching { context.dataDir }.getOrNull(),
        ).also { preferences -> preferences.publishIfStale() }
    }
}

File: apps/ios/TypviaKit/Sources/Core/UiPreferences.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

import Foundation
import SwiftUI

/// The two things a reader can set about how the product looks.
///
/// They are stored in the App Group rather than in the app's own defaults, so
/// a surface that is not the app can read them — a keyboard that ignores the
/// app's language setting is the mixed-language screen the design forbids,
/// arriving through the back door.
///
/// Storing them there is not the same as reading them there. Each surface has
/// to ask, at the moment it appears: the keyboard in its controller, the share
/// sheet as it builds its host, the widget as it builds a timeline entry. This
/// comment once claimed all of them while only the app read anything, which is
/// how a promise ends up in a file nobody rereads.
///
/// Nothing else lives here. These are UI preferences; anything about snippets
/// belongs to the database.
public enum UiPreference {
    /// Which language the interface is rendered in.
    public enum Language: String, CaseIterable, Sendable {
        /// Whatever the phone is set to.
        case system
        case en
        case zh

        /// Resolves to the one that actually gets rendered.
        public func resolved(preferred: [String] = Locale.preferredLanguages) -> UiLanguage {
            switch self {
            case .system: UiLanguage.resolve(preferred: preferred)
            case .en: .en
            case .zh: .zh
            }
        }
    }

    /// Light, dark, or the phone's own.
    public enum Appearance: String, CaseIterable, Sendable {
        case system
        case light
        case dark

        public var colorScheme: ColorScheme? {
            switch self {
            case .system: nil
            case .light: .light
            case .dark: .dark
            }
        }
    }
}

/// Reads and writes the two preferences.
///
/// An unset preference is `system`, and an unreadable store is also `system` —
/// the product works without ever being configured, and a failure to read a
/// preference must not stop it from starting.
@MainActor
public final class UiPreferences: ObservableObject {
    @Published public var language: UiPreference.Language {
        didSet { write(Keys.language, language.rawValue) }
    }

    @Published public var appearance: UiPreference.Appearance {
        didSet { write(Keys.appearance, appearance.rawValue) }
    }

    private let defaults: UserDefaults?

    public init(defaults: UserDefaults? = UserDefaults(suiteName: SharedContainer.appGroup)) {
        self.defaults = defaults
        let stored = UiPreferences.chosen(defaults: defaults)
        language = stored.language
        appearance = stored.appearance
    }

    private func write(_ key: String, _ value: String) {
        defaults?.set(value, forKey: key)
    }

    /// Reads the store again.
    ///
    /// For the surfaces that are not the app: an extension is a process of its
    /// own and is not told when the reader chooses something, so it has to ask
    /// again at the moment it comes up. Without this a keyboard built once
    /// keeps the language it was built with until the system happens to end
    /// the process.
    public func reload() {
        let stored = UiPreferences.chosen(defaults: defaults)
        if stored.language != language { language = stored.language }
        if stored.appearance != appearance { appearance = stored.appearance }
    }

    /// The two choices as they stand, read without an object to observe them.
    ///
    /// A widget is drawn from a value in a process that has no moment to hold
    /// an observable and nothing to observe it from; a share sheet builds its
    /// host once and is gone. They read here, so the parse of an unknown or
    /// missing value exists once rather than once per surface.
    public nonisolated static func chosen(
        defaults: UserDefaults? = UserDefaults(suiteName: SharedContainer.appGroup)
    ) -> Chosen {
        Chosen(
            language: UiPreference.Language(
                rawValue: defaults?.string(forKey: Keys.language) ?? ""
            ) ?? .system,
            appearance: UiPreference.Appearance(
                rawValue: defaults?.string(forKey: Keys.appearance) ?? ""
            ) ?? .system
        )
    }

    /// Both preferences as one value.
    public struct Chosen: Equatable, Sendable {
        public let language: UiPreference.Language
        public let appearance: UiPreference.Appearance

        public init(
            language: UiPreference.Language = .system,
            appearance: UiPreference.Appearance = .system
        ) {
            self.language = language
            self.appearance = appearance
        }
    }

    /// The keys are namespaced the way the desktop's are, so the two products
    /// read as one when someone goes looking.
    enum Keys {
        static let language = "tv.ui.locale"
        static let appearance = "tv.ui.theme"
    }
}

extension View {
    /// Applies the reader's choices to a whole tree: the language every copy
    /// pair renders in, and the appearance the paper takes.
    @MainActor
    public func uiPreferences(_ preferences: UiPreferences) -> some View {
        uiPreferences(
            UiPreferences.Chosen(
                language: preferences.language,
                appearance: preferences.appearance
            )
        )
    }

    /// The same, from a value rather than from something to observe — which is
    /// all a widget's face has.
    public func uiPreferences(_ chosen: UiPreferences.Chosen) -> some View {
        environment(\.tr, Translator(language: chosen.language.resolved()))
            .modifier(ChosenAppearance(appearance: chosen.appearance))
    }
}

/// The chosen appearance, applied at the two layers that read it.
///
/// The surfaces do not agree on which lever they honour: `preferredColorScheme`
/// reaches the window a screen sits in and does nothing in a widget, while the
/// environment value is what a dynamic colour resolves against. One choice, set
/// in both places — and set in neither when the choice is to follow the phone,
/// because an environment pinned to one scheme is not "system", it is that
/// scheme.
private struct ChosenAppearance: ViewModifier {
    let appearance: UiPreference.Appearance

    @ViewBuilder
    func body(content: Content) -> some View {
        if let scheme = appearance.colorScheme {
            content
                .environment(\.colorScheme, scheme)
                .preferredColorScheme(scheme)
        } else {
            content
        }
    }
}
